### PR TITLE
epoch fix

### PR DESCRIPTION
**Pull request by Transpeptidase**
This PR fixes a bug in LightEpoch, which caused by memory reorder.

Fix [#149](https://github.com/microsoft/FASTER/issues/149)

### CODE REVIEW

**Comment by badrishc** [comment]
Seems to be failing a few unit tests

**Comment by Transpeptidase** [comment]
> Seems to be failing a few unit tests

I don't know if it is running too slowly or deadlock.

**Comment by Transpeptidase** [comment]
> Seems to be failing a few unit tests

I fixed it.

File: cc/src/core/light_epoch.h [hunk 1]
```diff
@@ -176,44 +176,58 @@ class LightEpoch {
...
+
+    return e;
   }
 
   /// Enter the thread into the protected code region
   /// Process entries in drain list if possible
   inline uint64_t ProtectAndDrain() {
     uint32_t entry = Thread::id();
-    table_[entry].local_current_epoch = current_epoch.load();
+    auto e = current_epoch.load(std::memory_order_relaxed);
```
Comment (tli2) [inline cc/src/core/light_epoch.h]: These three lines of code appears duplicated multiple times throughout the file. Consider putting them into a function?

**Comment by Transpeptidase** [comment]
> First of all, thanks for finding the bug and submitting a PR!
> 
> If you have more time to work on this:
> 
> * The use of std::atomic mixed with fences seems overly complicated. By default every operation on atomic should operate under `std::memory_order_seq_cst` and should do the correct thing for you. Now it is conceivable that we lose some performance due to excessive synchronization, but let's back that up with benchmark results and not prematurely optimize.
> * Alternatively, consider using `std::atomic_signal_fence` without the need to turn `local_current_epoch` into an atomic.
> * Can you add your test code as a unit test?

Unfortunately, in C++ consistency model, ``memory_order_acquire`` only prevents the reorder of load-store and load-load. Thus, the ``store(memory_order_acquire)`` and ``std::atomic_thread_fence
(memory_order_acquire)`` don't work. ``memory_order_seq_cst`` is not always doing  the correct thing,  the ``store(memory_order_seq_cst)`` is equivalent to ``store(memory_order_acquire)``, which 
allows the reorder of  store-load. (see https://en.cppreference.com/w/cpp/atomic/memory_order,  and you can verify it  by generate assembly code with ``g++ -O3 -S``). 
I have not used `std::atomic_signal_fence` before. But  it seems to only prevent reorder by compiler rather than reorder by CPU (https://en.cppreference.com/w/cpp/atomic/atomic_signal_fence).

**Comment by tli2** [comment]
> First of all, thanks for finding the bug and submitting a PR!
> If you have more time to work on this:
> 
> The use of std::atomic mixed with fences seems overly complicated. By default every operation on atomic should operate under std::memory_order_seq_cst and should do the correct thing for you. Now it is conceivable that we lose some performance due to excessive synchronization, but let's back that up with benchmark results and not prematurely optimize.
> Alternatively, consider using std::atomic_signal_fence without the need to turn local_current_epoch into an atomic.
> Can you add your test code as a unit test?
> 
> 
> Unfortunately, in C++ consistency model, memory_order_acquire only prevents the reorder of load-store and load-load. Thus, the store(memory_order_acquire) and std::atomic_thread_fence (memory_order_acquire) don't work. memory_order_seq_cst is not always doing  the correct thing,  the store(memory_order_seq_cst) is equivalent to store(memory_order_acquire), which
> allows the reorder of  store-load. (see https://en.cppreference.com/w/cpp/atomic/memory_order,  and you can verify it  by generate assembly code with g++ -O3 -S).
> I have not used std::atomic_signal_fence before. But  it seems to only prevent reorder by compiler rather than reorder by CPU (https://en.cppreference.com/w/cpp/atomic/atomic_signal_fence).

Seems like it's worth a comment in the code in your PR.

**Comment by badrishc** [comment]
We have not yet seen this to be an issue that repros in FASTER. The matter requires further investigation. I wonder if it happens only on certain architectures. Closing for now, we can reopen if we get a clearer idea of the issue in terms of a FASTER testcase failure.

**Pull request closed without merging**